src/EditModel/event-program/program-access/ProgramStagesAccess.js

```diff
@@ -157,114 +154,101 @@ class ProgramStagesAccess extends React.Component {
-                programSharing: updatedSharing,
-            }, this.updateInternalState);
-    }
+        const programWithSharing = {
+            id: sharingDialogResult.id,
+            sharing: updatedSharing,
+        };
+        this.setState({
+            programSharing: programWithSharing,
+        });
+    };
 
-    updateStageAccess = updatedSharing => {
-        this.storeStageChanges(updatedSharing.id, updatedSharing);
+    updateStageAccess = (sharingDialogResult, updatedSharing) => {
+        this.storeStageChanges(sharingDialogResult.id, updatedSharing);
         this.setState({
-            stagesSharing: this.state.stagesSharing.map(stage =>
-                updatedSharing.id === stage.id
-                    ? updatedSharing
-                    : stage
+            stagesSharing: this.state.stagesSharing.map(
+                stage =>
+                    sharingDialogResult.id === stage.id
+                        ? { ...stage, sharing: updatedSharing }
+                        : stage
             ),
-        }, this.updateInternalState);
-    }
+        });
+    };
 
-    storeProgramChanges = (program) => {
-        sharingFields.forEach(property => {
-            this.props.editFieldChanged(
-                property,
-                program[property],
-            )
-        })
-    }
+    storeProgramChanges = sharingProperties => {
+        this.props.editFieldChanged('sharing', sharingProperties);
+    };
 
     storeStageChanges = (stageId, sharingProperties) => {
-        sharingFields.forEach(property => {
-            this.props.editProgramStageField(
-                stageId,
-                property,
-                sharingProperties[property],
-            );
-        });
-    }
+        this.props.editProgramStageField(stageId, 'sharing', sharingProperties);
+    };
 
     propagateAccess = event => {
         event.stopPropagation();
 
-        this.state.stagesSharing.forEach(stage => {
-            if (this.state.selectedStages.includes(stage.id)) {
-                this.storeStageChanges(stage.id, this.state.programSharing);
-            }
-        });
-
         const propagateIfSelected = stage => {
+            const programSharing = this.state.programSharing.sharing;
             if (this.state.selectedStages.includes(stage.id)) {
+                this.storeStageChanges(
+                    stage.id,
+                    programSharing
+                );
                 return {
-                    ...this.state.programSharing,
+                    sharing: programSharing,
                     id: stage.id,
                     displayName: stage.displayName || stage.name,
                 };
             } else return stage;
         };
-
-        this.setState(
-            {
-                stagesSharing: this.state.stagesSharing.map(propagateIfSelected),
-            },
-            this.updateInternalState,
-        );
-    };
-
-    updateInternalState = () => {
```

Comment:
This is a typical example where this should just be calculated instead of put in state. Only used in render.